File: src/examples/simple_room.ts
import { Vector3, Triangle, Light } from '../index';

/**
 * Creates a simple cubic room with a central block.
 * Room dimensions: 10×10×10 units
 * Block dimensions: 2×2×2 units at the center
 */
export function createSimpleRoomEnvironment(): {
  triangles: Triangle[];
  lights: Light[];
} {
  const triangles: Triangle[] = [];

  // Room dimensions
  const roomSize = 10;
  const roomMin = 0;
  const roomMax = roomSize;

  // Room construction: all triangles face inward
  // Each wall consists of 2 triangles

  // Floor (y = 0, faces upward into room)
  triangles.push(
    new Triangle(
      new Vector3(roomMin, roomMin, roomMin),
      new Vector3(roomMax, roomMin, roomMax),
      new Vector3(roomMin, roomMin, roomMax)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(roomMin, roomMin, roomMin),
      new Vector3(roomMax, roomMin, roomMin),
      new Vector3(roomMax, roomMin, roomMax)
    )
  );

  // Ceiling (y = roomMax, faces downward into room)
  triangles.push(
    new Triangle(
      new Vector3(roomMin, roomMax, roomMin),
      new Vector3(roomMin, roomMax, roomMax),
      new Vector3(roomMax, roomMax, roomMax)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(roomMin, roomMax, roomMin),
      new Vector3(roomMax, roomMax, roomMax),
      new Vector3(roomMax, roomMax, roomMin)
    )
  );

  // Front wall (z = roomMin, faces inward)
  triangles.push(
    new Triangle(
      new Vector3(roomMin, roomMin, roomMin),
      new Vector3(roomMin, roomMax, roomMin),
      new Vector3(roomMax, roomMax, roomMin)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(roomMin, roomMin, roomMin),
      new Vector3(roomMax, roomMax, roomMin),
      new Vector3(roomMax, roomMin, roomMin)
    )
  );

  // Back wall (z = roomMax, faces inward)
  triangles.push(
    new Triangle(
      new Vector3(roomMax, roomMin, roomMax),
      new Vector3(roomMax, roomMax, roomMax),
      new Vector3(roomMin, roomMax, roomMax)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(roomMax, roomMin, roomMax),
      new Vector3(roomMin, roomMax, roomMax),
      new Vector3(roomMin, roomMin, roomMax)
    )
  );

  // Left wall (x = roomMin, faces inward)
  triangles.push(
    new Triangle(
      new Vector3(roomMin, roomMin, roomMax),
      new Vector3(roomMin, roomMax, roomMax),
      new Vector3(roomMin, roomMax, roomMin)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(roomMin, roomMin, roomMax),
      new Vector3(roomMin, roomMax, roomMin),
      new Vector3(roomMin, roomMin, roomMin)
    )
  );

  // Right wall (x = roomMax, faces inward)
  triangles.push(
    new Triangle(
      new Vector3(roomMax, roomMin, roomMin),
      new Vector3(roomMax, roomMax, roomMin),
      new Vector3(roomMax, roomMax, roomMax)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(roomMax, roomMin, roomMin),
      new Vector3(roomMax, roomMax, roomMax),
      new Vector3(roomMax, roomMin, roomMax)
    )
  );

  // Central hovering block (2×2×2, centered at 5,4,5)
  // Triangles face outward
  const blockMin = 4;
  const blockMax = 6;

  // Block bottom (y = blockMin, faces downward/outward)
  triangles.push(
    new Triangle(
      new Vector3(blockMin, blockMin, blockMin),
      new Vector3(blockMin, blockMin, blockMax),
      new Vector3(blockMax, blockMin, blockMax)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(blockMin, blockMin, blockMin),
      new Vector3(blockMax, blockMin, blockMax),
      new Vector3(blockMax, blockMin, blockMin)
    )
  );

  // Block top (y = blockMax, faces upward/outward)
  triangles.push(
    new Triangle(
      new Vector3(blockMin, blockMax, blockMin),
      new Vector3(blockMax, blockMax, blockMax),
      new Vector3(blockMin, blockMax, blockMax)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(blockMin, blockMax, blockMin),
      new Vector3(blockMax, blockMax, blockMin),
      new Vector3(blockMax, blockMax, blockMax)
    )
  );

  // Block front (z = blockMin, faces outward)
  triangles.push(
    new Triangle(
      new Vector3(blockMin, blockMin, blockMin),
      new Vector3(blockMax, blockMin, blockMin),
      new Vector3(blockMax, blockMax, blockMin)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(blockMin, blockMin, blockMin),
      new Vector3(blockMax, blockMax, blockMin),
      new Vector3(blockMin, blockMax, blockMin)
    )
  );

  // Block back (z = blockMax, faces outward)
  triangles.push(
    new Triangle(
      new Vector3(blockMin, blockMin, blockMax),
      new Vector3(blockMax, blockMax, blockMax),
      new Vector3(blockMax, blockMin, blockMax)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(blockMin, blockMin, blockMax),
      new Vector3(blockMin, blockMax, blockMax),
      new Vector3(blockMax, blockMax, blockMax)
    )
  );

  // Block left (x = blockMin, faces outward)
  triangles.push(
    new Triangle(
      new Vector3(blockMin, blockMin, blockMin),
      new Vector3(blockMin, blockMax, blockMin),
      new Vector3(blockMin, blockMax, blockMax)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(blockMin, blockMin, blockMin),
      new Vector3(blockMin, blockMax, blockMax),
      new Vector3(blockMin, blockMin, blockMax)
    )
  );

  // Block right (x = blockMax, faces outward)
  triangles.push(
    new Triangle(
      new Vector3(blockMax, blockMin, blockMin),
      new Vector3(blockMax, blockMax, blockMax),
      new Vector3(blockMax, blockMax, blockMin)
    )
  );
  triangles.push(
    new Triangle(
      new Vector3(blockMax, blockMin, blockMin),
      new Vector3(blockMax, blockMin, blockMax),
      new Vector3(blockMax, blockMax, blockMax)
    )
  );

  // Single light at top center of room, just below ceiling
  const lights = [new Light(new Vector3(5, 9.5, 5), 1000)]; // 1000 W intensity

  return { triangles, lights };
}

/**
 * Generate test points on the floor of the room at various positions
 */
export function generateFloorTestPoints(): Vector3[] {
  const points: Vector3[] = [];
  const floorY = 0.1; // Slightly above the floor

  // Create a grid of test points
  for (let x = 1; x < 10; x += 2) {
    for (let z = 1; z < 10; z += 2) {
      points.push(new Vector3(x, floorY, z));
    }
  }

  return points;
}

File: src/examples/simple_room.py
"""Simple example: cubic room with central block"""

from typing import List, Tuple
from ..core import Vector3, Triangle, Light


def create_simple_room_environment() -> Tuple[List[Triangle], List[Light]]:
    """
    Creates a simple cubic room with a central block.
    Room dimensions: 10×10×10 units
    Block dimensions: 2×2×2 units at the center
    """
    triangles: List[Triangle] = []

    # Room dimensions
    room_size = 10
    room_min = 0
    room_max = room_size

    # Room construction: all triangles face inward
    # Each wall consists of 2 triangles

    # Floor (y = 0, faces upward into room)
    triangles.append(
        Triangle(
            Vector3(room_min, room_min, room_min),
            Vector3(room_max, room_min, room_max),
            Vector3(room_min, room_min, room_max),
        )
    )
    triangles.append(
        Triangle(
            Vector3(room_min, room_min, room_min),
            Vector3(room_max, room_min, room_min),
            Vector3(room_max, room_min, room_max),
        )
    )

    # Ceiling (y = room_max, faces downward into room)
    triangles.append(
        Triangle(
            Vector3(room_min, room_max, room_min),
            Vector3(room_min, room_max, room_max),
            Vector3(room_max, room_max, room_max),
        )
    )
    triangles.append(
        Triangle(
            Vector3(room_min, room_max, room_min),
            Vector3(room_max, room_max, room_max),
            Vector3(room_max, room_max, room_min),
        )
    )

    # Front wall (z = room_min, faces inward)
    triangles.append(
        Triangle(
            Vector3(room_min, room_min, room_min),
            Vector3(room_min, room_max, room_min),
            Vector3(room_max, room_max, room_min),
        )
    )
    triangles.append(
        Triangle(
            Vector3(room_min, room_min, room_min),
            Vector3(room_max, room_max, room_min),
            Vector3(room_max, room_min, room_min),
        )
    )

    # Back wall (z = room_max, faces inward)
    triangles.append(
        Triangle(
            Vector3(room_max, room_min, room_max),
            Vector3(room_max, room_max, room_max),
            Vector3(room_min, room_max, room_max),
        )
    )
    triangles.append(
        Triangle(
            Vector3(room_max, room_min, room_max),
            Vector3(room_min, room_max, room_max),
            Vector3(room_min, room_min, room_max),
        )
    )

    # Left wall (x = room_min, faces inward)
    triangles.append(
        Triangle(
            Vector3(room_min, room_min, room_max),
            Vector3(room_min, room_max, room_max),
            Vector3(room_min, room_max, room_min),
        )
    )
    triangles.append(
        Triangle(
            Vector3(room_min, room_min, room_max),
            Vector3(room_min, room_max, room_min),
            Vector3(room_min, room_min, room_min),
        )
    )

    # Right wall (x = room_max, faces inward)
    triangles.append(
        Triangle(
            Vector3(room_max, room_min, room_min),
            Vector3(room_max, room_max, room_min),
            Vector3(room_max, room_max, room_max),
        )
    )
    triangles.append(
        Triangle(
            Vector3(room_max, room_min, room_min),
            Vector3(room_max, room_max, room_max),
            Vector3(room_max, room_min, room_max),
        )
    )

    # Central hovering block (2×2×2, centered at 5,4,5)
    # Triangles face outward
    block_min = 4
    block_max = 6

    # Block bottom (y = block_min, faces downward/outward)
    triangles.append(
        Triangle(
            Vector3(block_min, block_min, block_min),
            Vector3(block_min, block_min, block_max),
            Vector3(block_max, block_min, block_max),
        )
    )
    triangles.append(
        Triangle(
            Vector3(block_min, block_min, block_min),
            Vector3(block_max, block_min, block_max),
            Vector3(block_max, block_min, block_min),
        )
    )

    # Block top (y = block_max, faces upward/outward)
    triangles.append(
        Triangle(
            Vector3(block_min, block_max, block_min),
            Vector3(block_max, block_max, block_max),
            Vector3(block_min, block_max, block_max),
        )
    )
    triangles.append(
        Triangle(
            Vector3(block_min, block_max, block_min),
            Vector3(block_max, block_max, block_min),
            Vector3(block_max, block_max, block_max),
        )
    )

    # Block front (z = block_min, faces outward)
    triangles.append(
        Triangle(
            Vector3(block_min, block_min, block_min),
            Vector3(block_max, block_min, block_min),
            Vector3(block_max, block_max, block_min),
        )
    )
    triangles.append(
        Triangle(
            Vector3(block_min, block_min, block_min),
            Vector3(block_max, block_max, block_min),
            Vector3(block_min, block_max, block_min),
        )
    )

    # Block back (z = block_max, faces outward)
    triangles.append(
        Triangle(
            Vector3(block_min, block_min, block_max),
            Vector3(block_max, block_max, block_max),
            Vector3(block_max, block_min, block_max),
        )
    )
    triangles.append(
        Triangle(
            Vector3(block_min, block_min, block_max),
            Vector3(block_min, block_max, block_max),
            Vector3(block_max, block_max, block_max),
        )
    )

    # Block left (x = block_min, faces outward)
    triangles.append(
        Triangle(
            Vector3(block_min, block_min, block_min),
            Vector3(block_min, block_max, block_min),
            Vector3(block_min, block_max, block_max),
        )
    )
    triangles.append(
        Triangle(
            Vector3(block_min, block_min, block_min),
            Vector3(block_min, block_max, block_max),
            Vector3(block_min, block_min, block_max),
        )
    )

    # Block right (x = block_max, faces outward)
    triangles.append(
        Triangle(
            Vector3(block_max, block_min, block_min),
            Vector3(block_max, block_max, block_max),
            Vector3(block_max, block_max, block_min),
        )
    )
    triangles.append(
        Triangle(
            Vector3(block_max, block_min, block_min),
            Vector3(block_max, block_min, block_max),
            Vector3(block_max, block_max, block_max),
        )
    )

    # Single light at top center of room, just below ceiling
    lights = [Light(Vector3(5, 9.5, 5), 1000)]  # 1000 W intensity

    return triangles, lights


def generate_floor_test_points() -> List[Vector3]:
    """Generate test points on the floor of the room at various positions"""
    points: List[Vector3] = []
    floor_y = 0.1  # Slightly above the floor

    # Create a grid of test points
    for x in range(1, 20, 1):
        for z in range(1, 20, 1):
            points.append(Vector3(x/2, floor_y, z/2))

    return points
